nucypher/blockchain/eth/sol/source/contracts/lib/Snapshot.sol

```diff
@@ -0,0 +1,104 @@
+pragma solidity ^0.6.5;
+
+/**
+ * @title Snapshot
+ * @notice Manages snapshots of size 128 bits (32 bits for timestamp, 96 bits for value)
+ * 96 bits is enough for storing NU token values, and 32 bits should be OK for block numbers
+ * @dev Since each storage slot can hold two snapshots, new slots are allocated every other TX. Thus, gas cost of adding snapshots is 51400 and 36400 gas, alternately.
+ * Based on Aragon's Checkpointing (https://https://github.com/aragonone/voting-connectors/blob/master/shared/contract-utils/contracts/Checkpointing.sol)
+ * On average, adding snapshots spends ~6500 less gas than the 256-bit checkpoints of Aragon's Checkpointing
+ */
+library Snapshot {
+
+    function encodeSnapshot(uint32 _time, uint96 _value) internal pure returns(uint128) {
+        return uint128(uint256(_time) << 96 | uint256(_value));
+    }
+
+    function decodeSnapshot(uint128 _snapshot) internal pure returns(uint32 time, uint96 value){
+        time = uint32(bytes4(bytes16(_snapshot)));
+        value = uint96(_snapshot);
+    }
+
+    function addSnapshot(uint128[] storage _self, uint256 _value) internal {
+        addSnapshot(_self, block.number, _value);
+    }
+
+    function addSnapshot(uint128[] storage _self, uint256 _time, uint256 _value) internal {
+        uint256 length = _self.length;
+        if (length != 0) {
+            (uint32 currentTime, uint96 currentValue) = decodeSnapshot(_self[length - 1]);
+            if (uint32(_time) == currentTime) {
+                _self[length - 1] = encodeSnapshot(uint32(_time), uint96(_value));
+                return;
+            } else if (uint32(_time) < currentTime){
+                revert();
+            }
+        }
+        _self.push(encodeSnapshot(uint32(_time), uint96(_value)));
+    }
+
+    function lastSnapshot(uint128[] storage _self) internal view returns (uint32, uint96) {
+        uint256 length = _self.length;
+        if (length > 0) {
+            return decodeSnapshot(_self[length - 1]);
+        }
+
+        return (0, 0);
+    }
+
+    function lastValue(uint128[] storage _self) internal view returns (uint96) {
```

Comment:
if return type will be `uint256` than will be fewer casts and I think code size will be smaller